### ISSUE TITLE
memory corruption in python bindings?

### ISSUE THREAD

**Issue by sigmonsays** [acg/lwpb#2]
Symptoms are seen on the 'python' branch using the python bindings. There seems to be a very strange type of error where the python stack, memory or state in general is getting smashed up. Please see the following example at  https://gist.github.com/11c6f1b509770b848a52 as github markup formatting couldn't easily preformat blocks of text.

Line #30 from the gist explodes. The output looks like this:
$ python test.py 
1
2
Traceback (most recent call last):
  File "test_perf_lwpb.py", line 13, in <module>
    for n in [1,2]:
OverflowError: could not convert to long

Platform is ubuntu lucid

 dpkg -l |grep protobuf
ii  libprotobuf-c0                    0.14-1                                          Protocol Buffers C library
ii  libprotobuf-c0-dev                0.14-1                                          Protocol Buffers C static library and header
ii  libprotobuf-dev                   2.4.0a-2ubuntu2                                 protocol buffers C++ library (development fi
ii  libprotobuf-lite7                 2.4.0a-2ubuntu2                                 protocol buffers C++ library (lite version)
rc  libprotobuf5                      2.2.0a-0.1ubuntu1                               protocol buffers C++ library
ii  libprotobuf7                      2.4.0a-2ubuntu2                                 protocol buffers C++ library
ii  protobuf-c-compiler               0.14-1                                          protocol buffers C compiler
ii  protobuf-compiler                 2.4.0a-2ubuntu2                                 compiler for protocol buffer definition file
ii  python-protobuf                   2.4.0a-2ubuntu2                                 Python bindings for protocol buffers

**Comment by acg** [comment]
Reproducible. Caused by setting a default value for an enum field. If you take out the default value things are fine. I avoid both defaults and enums generally, but it would be nice if this worked, and it DEFINITELY shouldn't cause a corruption as you're seeing currently.

I'll try to get around to fixing this, but if you get impatient, the place to look would be in the .pb2 file description structure in lwpb/__ init __.py and the code in descriptor.c around line 226.

-Alan

**Comment by sigmonsays** [comment]
thanks for the confirmation. I too am tight on time (aren't we all). I'll let ya know if I make a fix so we can get it upstream.

**Comment by acg** [comment]
I just pushed a fix for the immediate problem you were seeing, which turned out to not really be a corruption (see below). However in looking through the code again I realized the C lwpb library, by design, doesn't apply default values or enforce required fields -- see [limitations](http://code.google.com/p/lwpb/#Limitations). So if you're actually using these features of protobuf, there's more work to do. I'm opening a separate issue for this.

The "corruption": it turns out protoc does _not_ encode default values as the corresponding primitive type in the compiled .pb2 descriptor file. It encodes them as ascii strings. So the descriptor code needs to always expect a string, and then call something like sscanf() to create a default value. It's actually worse, because default values for enums are not the numeric value, but rather a symbol string ("OP_TWO" in your example) that need to be looked up in an enum table elsewhere, subject to C++ style scoping resolution, agg blarg barf barf...

To make a long story short, there was an error converting the string "OP_TWO" into a long value for the enum. The error didn't propagate immediately but was stored in Python's version of errno. Some later python code actually checked it and propagated the long conversion exception, resulting in the exception you saw.

The new code creates default values or throws an exception immediately if it can't. Like I said though, neither the lwpb encoder nor the decoder actually apply default values currently.
